Limit camera movement within explored region

Gualdimar opened the pull request:
A rectangular area is calculated from the two extreme explored points on the map, beyond which the camera cannot go.
The area is recalculated after each new tile is explored and stored for each civilization separately, so they all have their own area.
Since the hex coordinate system is offset by an angle relative to the world one, longitude and latitude were used instead of hex coordinates to calculate the extreme points. The longitude and latitude are then converted to the world coords. This allowed the use of two points on the map, instead of four when using hex coordinates.

vegeta1k95 commented:
This breaks world-wrap continious scrolling.
Create world-wrap world with whole map explored (start at Future Era): scrolling is stuck at map boundaries.

vegeta1k95 commented on core/src/com/unciv/logic/civilization/Civilization.kt:
```diff
@@ -191,6 +196,7 @@ class Civilization : IsPartOfGameInfoSerialization {
     var cities = listOf<City>()
     var citiesCreated = 0
     var exploredTiles = HashSet<Vector2>()
+    var exploredRegion = HashMap<String,Vector2>()
```
Since you made this a non-transient serializable field, there should be a correspondent line in the `Civilization.clone()`

vegeta1k95 commented on core/src/com/unciv/ui/utils/ZoomableScrollPane.kt:
```diff
@@ -71,13 +71,35 @@ open class ZoomableScrollPane(
 
     override fun scrollX(pixelsX: Float) {
-        super.scrollX(pixelsX)
+        val exploredRegion = UncivGame.Current.worldScreen?.viewingCiv?.exploredRegion
+        var result = pixelsX
+        if(exploredRegion != null && !exploredRegion.isEmpty()) {
+            val centerX = this.maxX * 0.5f + exploredRegion["tileRadius"]!!.x
+            var min = centerX+exploredRegion["xWorldVector"]!!.x
+            val max = centerX+exploredRegion["xWorldVector"]!!.y
+
+            if (result < min) result = min
+            else if (result > max) result = max
+        }
+
+        super.scrollX(result)
         updateCulling()
         onViewportChanged()
     }
 
     override fun scrollY(pixelsY: Float) {
-        super.scrollY(pixelsY)
+        val exploredRegion = UncivGame.Current.worldScreen?.viewingCiv?.exploredRegion
+        var result = pixelsY
+        if(exploredRegion != null && !exploredRegion.isEmpty()) {
+            val centerY = this.maxY * 0.5f
+            var min = centerY-exploredRegion["yWorldVector"]!!.x
+            val max = centerY-exploredRegion["yWorldVector"]!!.y
+
+            if (result < min) result = min
+            else if (result > max) result = max
+        }
+
+        super.scrollY(result)
         updateCulling()
         onViewportChanged()
     }
```
I think this doesn't belong to `ZoomableScrollPane`, which is an abstract scroll pane object which in principle can contain everything, not only tile map. 

Moreover: CityScreen and MapEditorScreen which contain such scroll pane do not need camera limiting functionality: city screen has map size of like 8 tiles or so and map editor screen is not supposed to restrict vision of a map creator.

I'd propose to move this to `WorldMapHolder`, which is subclass of `ZoomableScrollPane` and already contains a reference to the current `WorldScreen`.

Gualdimar commented:
> This breaks world-wrap continious scrolling.

hm, world-wrap... i forgot about it. need to test.

Gualdimar commented:
@vegeta1k95 all your points are valid and can be easily resolved but man, the world wrap... the world wrap just makes things so complicated. I more or less got it to work, but the logic is so confusing, and the number of if statements is so large :exploding_head:
unresolved issues:
1. when 1-2 columns of FoW tiles are left unexplored, if you swing hard, you can jump over the limit and get to the other side
2. when exploring the last 2-4 columns of FoW tiles, it is difficult to determine from which side of the explored area to add them
3. tiles rendered before the edges of the explored area are connected. therefore, so you can see that a few FoW tiles away is a map explored for the other side

Gualdimar commented:
should now work correctly with world-wrap. better coordinate checking(there should be no overjumps).
new explored tiles are now added from the side closest to the explorer unit or just to the nearest side, if there is no explorer.

the last thing left is tile render over FoW

yairm210 commented on core/src/com/unciv/logic/civilization/Civilization.kt:
```diff
@@ -191,6 +197,7 @@ class Civilization : IsPartOfGameInfoSerialization {
     var cities = listOf<City>()
     var citiesCreated = 0
     var exploredTiles = HashSet<Vector2>()
+    var exploredRegion = HashMap<String,Vector2>()
```
Since there are only 2 vectors , this should not be a dictionary, treat them by name.
Also, call them 'topright' and 'bottomleft', 'start' and 'end' is very ambiguous

yairm210 replied:
Also, *this can be calculated from tilemap data*, so we don't need to persist this at all! All these can be transient!
(Since there are several vectors you can store them all in an object, like 'class exploredRegion')

Gualdimar replied:
> Also, _this can be calculated from tilemap data_, so we don't need to persist this at all! All these can be transient!

tested it, tiles become explored in a random order, sometimes messing up the explored region.

yairm210 commented on core/src/com/unciv/ui/utils/ZoomableScrollPane.kt:
```diff
@@ -366,4 +366,7 @@ open class ZoomableScrollPane(
         onViewportChangedListener?.invoke(maxX, maxY, getViewport())
     }
+
+    open fun checkX(deltaX: Float): Float = scrollX - deltaX
```
These are not good names. What is the intent for these functions?

Gualdimar replied:
These functions check if the coordinates are within the explored region

yairm210 replied:
So maybe getNewX? If a function is named "check" I expect to receive the result of a check, not the final information

yairm210 commented on core/src/com/unciv/ui/worldscreen/WorldMapHolder.kt:
```diff
@@ -754,4 +754,37 @@ class WorldMapHolder(
+    override fun checkY(deltaY: Float): Float {
+        val exploredRegion = worldScreen.viewingCiv.exploredRegion
+        var result = scrollY + deltaY
+        if(!exploredRegion.isEmpty()) {
```
Reverse this if. Also, under what circumstances is the exploredregion empty? How can that be?

Gualdimar replied:
If acts as a safeguard if something goes wrong so that the camera doesn't get stuck at the wrong point. I'm not sure I can think through all the possible scenarios.

yairm210 commented on core/src/com/unciv/ui/worldscreen/WorldMapHolder.kt:
```diff
@@ -754,4 +754,37 @@ class WorldMapHolder(
+        var result = scrollY + deltaY
+        if(!exploredRegion.isEmpty()) {
+            val mapCenterY = maxY * 0.5f
+            val startY = mapCenterY-exploredRegion["yWorldVector"]!!.x
```
This is another confusing thing. You're using x,y in the vector for usages that are not, in fact, x and y.
This would be better served by topright/bottomleft vectors for world, OR just use maxWorldY, minWorldY etc in the exploredRegion object

yairm210 commented on core/src/com/unciv/logic/civilization/Civilization.kt:
```diff
@@ -802,6 +810,80 @@ class Civilization : IsPartOfGameInfoSerialization {
 
     fun asPreview() = CivilizationInfoPreview(this)
+
+    fun calculateExploredRegion(position: Vector2, explorerPosition: Vector2?) {
```
I hope after the existing comments I gave this will be more readable, but I still think this should definitely have more comments. You seem to be dealing with edge cases for world wrap that are unclear when reading the code, so just saying in lines like 'we passed over the edge of the world' could make things much clearer

yairm210 replied:
(also why do you need the explorer position?)

Gualdimar replied:
ok, I will definitely add comments. I was wondering what you would say about this new feature in general, whether it deserves the right to life)

when exploring the last 2-4 columns of FoW tiles, it is difficult to determine from which side of the explored area to add them, because the new cell can be right in the middle between them. so I use the position of the explorer to determine from which side to expand the region

Gualdimar commented:
Updated:
- Loading the game from the loading screen resets the current scroll position and zoom. Why: When loading a previously saved game, the camera may have been outside the explored region.
- Moved all logic to a separate class ExploredRegion
- Reduced the number of non-transient properties
- Added comments

Still have a problem with the visibility of the other edge of the map through the fog. I haven't come up with a good solution.
![24_1](https://user-images.githubusercontent.com/1225948/220873298-9e1f5ee5-a7a4-43b9-b084-8f530158e1e5.jpg)

yairm210 commented:
I think that's fine, this solves 95% of the problem :)

yairm210 commented:
Now all we need is 'minimap displays only explored area' to have truly location-agnistic non-world-wrap games :) (world wrap ones would also need to account for "non-centered map when crossing edge")
